edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/CloudConnectionProvider.cs

```diff
@@ -109,31 +116,95 @@ internal static ITransportSettings[] GetTransportSettings(Option<UpstreamProtoco
                         });
         }
 
-        public async Task<Try<ICloudConnection>> Connect(IClientCredentials identity, Action<string, CloudConnectionStatus> connectionStatusChangedHandler)
+        public async Task<Try<ICloudConnection>> Connect(IClientCredentials clientCredentials, Action<string, CloudConnectionStatus> connectionStatusChangedHandler)
+        {
+            Preconditions.CheckNotNull(clientCredentials, nameof(clientCredentials));
+
+            try
+            {
+                Events.CreatingCloudConnectionUsingClientCredentials(clientCredentials);
+                var cloudListener = new CloudListener(this.edgeHub.Expect(() => new InvalidOperationException("EdgeHub reference should not be null")), clientCredentials.Identity.Id);
+
+                if (this.edgeHubIdentity.Id.Equals(clientCredentials.Identity.Id))
+                {
+                    ICloudConnection cc = await CloudConnection.Create(
+                        clientCredentials.Identity,
+                        connectionStatusChangedHandler,
+                        this.transportSettings,
+                        this.messageConverterProvider,
+                        this.clientProvider,
+                        cloudListener,
+                        this.edgeHubTokenProvider,
+                        this.idleTimeout,
+                        this.closeOnIdleTimeout);
+                    Events.SuccessCreatingCloudConnection(clientCredentials.Identity);
+                    return Try.Success(cc);
+                }
+                else if (clientCredentials is ITokenCredentials clientTokenCredentails)
+                {
+                    ICloudConnection cc = await ClientTokenCloudConnection.Create(
+                        clientTokenCredentails,
+                        connectionStatusChangedHandler,
+                        this.transportSettings,
+                        this.messageConverterProvider,
+                        this.clientProvider,
+                        cloudListener,
+                        this.idleTimeout,
+                        this.closeOnIdleTimeout);
+                    Events.SuccessCreatingCloudConnection(clientCredentials.Identity);
+                    return Try.Success(cc);
+                }
+                else
+                {
+                    throw new InvalidOperationException($"Cannot connect using client credentials of type {clientCredentials.AuthenticationType} for identity {clientCredentials.Identity.Id}");
```

Comment:
This is caught in the next line, maybe just return Try<ICloudConnection>.Failure instead of throwing

Reply:
Emm, I will have to duplicate the logging and throwing code. Maybe just keep it like this to keep the code clean? We don't expect to hit it often, if at all, so perf shouldn't be a big concern, I think?